drivers/adc/adc_ads1119.c

```diff
@@ -0,0 +1,522 @@
+		return rc;
+	}
+
+	rc = ads1119_wait_data_ready(dev);
+	if (rc != 0) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+
+	rc = ads1119_read_sample(dev, data->buffer);
+	if (rc != 0) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+	data->buffer++;
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	return rc;
+}
+
+#if CONFIG_ADC_ASYNC
+static int ads1119_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	rc = ads1119_adc_start_read(dev, sequence, true);
+	adc_context_release(&data->ctx, rc);
+
+	return rc;
+}
+#else
+static int ads1119_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	rc = ads1119_adc_start_read(dev, sequence, false);
+
+
+	while (rc == 0 && k_sem_count_get(&data->ctx.sync) == 0) {
+		rc = ads1119_adc_perform_read(dev);
+	}
+	k_sem_take(&data->ctx.sync, K_FOREVER);
+	adc_context_release(&data->ctx, rc);
+	return rc;
+}
+#endif
+
+#if CONFIG_ADC_ASYNC
+static void ads1119_acquisition_thread(struct device *dev)
+{
+	while (true) {
+		ads1119_adc_perform_read(dev);
+	}
+}
+#endif
+
+static int ads1119_init(const struct device *dev)
+{
+	int rc;
+	uint8_t status;
+	const struct ads1119_config *config = dev->config;
+	struct ads1119_data *data = dev->data;
+
+	adc_context_init(&data->ctx);
+
+	k_sem_init(&data->acq_sem, 0, 1);
+
+	if (!device_is_ready(config->i2c_dev)) {
+		return -ENODEV;
+	}
+
+	rc = ads1119_read_reg(dev, ADS1119_REG_STATUS, &status);
+	if (rc) {
+		LOG_ERR("Could not get %s ID", dev->name);
+		return rc;
+	}
+
+#if CONFIG_ADC_ASYNC
+	const k_tid_t tid =
+		k_thread_create(&data->thread, config->stack,
+				CONFIG_ADC_ADS1119_ACQUISITION_THREAD_STACK_SIZE,
+				(k_thread_entry_t)ads1119_acquisition_thread,
+				(void *)dev, NULL, NULL,
+				CONFIG_ADC_ADS1119_ASYNC_THREAD_INIT_PRIO,
+				0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_ads1119");
+#endif
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return rc;
+}
+
+static const struct adc_driver_api api = {
+	.channel_setup = ads1119_channel_setup,
+	.read = ads1119_read,
+	.ref_internal = ADS1119_REF_INTERNAL,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = ads1119_adc_read_async,
+#endif
+};
+#define ADC_ADS1119_INST_DEFINE(n)								 \
+	COND_CODE_1(CONFIG_ADC_ASYNC,								 \
+		    (static									 \
+		     K_KERNEL_STACK_DEFINE(thread_stack_##n,					 \
+					   CONFIG_ADC_ADS1119_ACQUISITION_THREAD_STACK_SIZE);),	 \
+		    ())										 \
+	static const struct ads1119_config config_##n = {					 \
+		.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(n)),					 \
+		.i2c_dt_device_addr = DT_INST_REG_ADDR(n),					 \
+		COND_CODE_1(CONFIG_ADC_ASYNC,							 \
+			    (.stack = thread_stack_##n),					 \
+			    ())									 \
+	};											 \
+	static struct ads1119_data data_##n;							 \
```

Comment:
the timer, lock, and sync should be initialized here, remove `adc_context_init` function
https://github.com/zephyrproject-rtos/zephyr/blob/64af112d30cd7340f2a32a4eaebe86cd995bdd06/drivers/adc/adc_ads1x1x.c#L620

Reply:
Doing so would mean that the whole ads1119_data structure will need to be allocated in the data section, which might be a waste of ROM. 
This is especially true when the whole structure is much bigger that the element that need to be initialized with something else than 0.
I think the ads1x1x driver would greatly benefit to not be statically initialized since the stack of the thread is a member of the structure. Meaning than a part of the ROM is only used to store the zeros to initialize the whole stack

Reply:
While I understand your point...that could be a discussion for a separate PR. This is the way _all_ ADCs handle it for now. not just the ads1x1x. It should be done within the data for now.
Perhaps @anangl can comment on why it's done this way.

Reply:
I will wait @anangl to comment on that

Reply:
There is no requirement to initialize the timer, lock, and async fields statically. I agree with @kurddt. It makes sense to initialize them at runtime when the data structure is big and all other fields can use the default initialization to 0.